File: tests/src/test/javascript/util/RandomTest.js
/**
 * Copyright (c) 2013-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Random unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("Random", function() {
    var Random = require('msl-core/util/Random.js');
    
    var ITERATIONS = 100000;
    var random = new Random();
    
    it("boolean", function() {
        var diff = 0;
        for (var i = 0; i < ITERATIONS; ++i)
            diff += (random.nextBoolean()) ? 1 : -1;
        expect(Math.abs(diff)).toBeLessThan(ITERATIONS * 0.01);
    });
    
    it("int" , function() {
        var positive = 0,
            negative = 0,
            xor = 0;
        for (var i = 0; i < ITERATIONS; ++i) {
            var val = random.nextInt();
            xor ^= val;
            (val > 0) ? ++positive : ++negative;
        }
        expect(positive).toBeGreaterThan(ITERATIONS * 0.49);
        expect(negative).toBeGreaterThan(ITERATIONS * 0.49);
        expect(positive - negative).toBeLessThan(ITERATIONS * 0.01);
        var diff = 0;
        for (var j = 0; j < 32; ++j)
            diff += (xor & (1 << j)) ? 1 : -1;
        expect(Math.abs(diff)).toBeLessThan(8);
    });
    
    it("int(n)", function() {
        var buckets = 10;
        var distribution = [];
        for (var b = 0; b < buckets; ++b)
            distribution[b] = 0;
        for (var i = 0; i < ITERATIONS; ++i) {
            var val = random.nextInt(buckets);
            ++distribution[val];
        }
        var average = ITERATIONS / buckets;
        for (var j = 0; j < buckets; ++j) {
            expect(distribution[j]).toBeGreaterThan(average * 0.9);
            expect(distribution[j]).toBeLessThan(average * 1.1);
        }
    });
    
    it("long", function() {
        var positive = 0,
        negative = 0,
        xor = 0;
        for (var i = 0; i < ITERATIONS; ++i) {
            var val = random.nextLong();
            xor ^= val;
            (val > 0) ? ++positive : ++negative;
        }
        expect(positive).toBeGreaterThan(ITERATIONS * 0.49);
        expect(negative).toBeGreaterThan(ITERATIONS * 0.49);
        expect(positive - negative).toBeLessThan(ITERATIONS * 0.01);
        var diff = 0;
        for (var j = 0; j < 32; ++j)
            diff += (xor & (1 << j)) ? 1 : -1;
        expect(Math.abs(diff)).toBeLessThan(8);
    });
    
    it("bytes", function() {
        var b = new Uint8Array(65536);
        random.nextBytes(b);
        
        var xor = 0;
        for (var i = 0; i < b.length; ++i) {
            xor ^= b[i];
        }
        var diff = 0;
        for (var j = 0; j < 8; ++j)
            diff += (xor & (1 << j)) ? 1 : -1;
        expect(Math.abs(diff)).toBeLessThan(3);
    });
    
    it("large bytes", function() {
        var b = new Uint8Array(2.5 * 65536);
        random.nextBytes(b);
        
        var xor = 0;
        for (var i = 0; i < b.length; ++i) {
            xor ^= b[i];
        }
        var diff = 0;
        for (var j = 0; j < 8; ++j)
            diff += (xor & (1 << j)) ? 1 : -1;
        expect(Math.abs(diff)).toBeLessThan(3);
    });
});

File: tests/src/test/cpp/crypto/RandomTest.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <crypto/OpenSslLib.h>
#include <crypto/Random.h>
#include <MslConstants.h>
#include <MslInternalException.h>
#include <math.h>
#include <chrono>
#include <thread>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <iomanip>
#include <set>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::crypto;

namespace netflix {
namespace msl {
namespace crypto {

namespace {

// http://www.johndcook.com/blog/standard_deviation/
class RunningStat
{
public:
    RunningStat() : m_n(0), m_oldM(0.), m_newM(0.), m_oldS(0.), m_newS(0.) {}
    void clear() { m_n = 0; }
    void push(double x) {
        m_n++;
        // See Knuth TAOCP vol 2, 3rd edition, page 232
        if (m_n == 1) {
            m_oldM = m_newM = x;
            m_oldS = 0.0;
        } else {
            m_newM = m_oldM + (x - m_oldM)/m_n;
            m_newS = m_oldS + (x - m_oldM)*(x - m_newM);

            // set up for next iteration
            m_oldM = m_newM;
            m_oldS = m_newS;
        }
    }
    int numDataValues() const { return m_n; }
    double mean() const { return (m_n > 0) ? m_newM : 0.0; }
    double variance() const {
        return ( (m_n > 1) ? m_newS/(m_n - 1) : 0.0 );
    }
    double standardDeviation() const { return sqrt( variance() ); }
private:
    int m_n;
    double m_oldM, m_newM, m_oldS, m_newS;
};

}

class RandomTest : public ::testing::Test
{
public:
    RandomTest() : random(make_shared<Random>()) {}
protected:
    shared_ptr<IRandom> random;
};

// http://www.johndcook.com/Beautiful_Testing_ch10.pdf
TEST_F(RandomTest, nextIntStats)
{
    RunningStat rs;
    const int N = 1000000;
    for (int i=0; i<N; ++i)
        rs.push(random->nextInt());

    // The range of an int32_t is -2,147,483,648..2,147,483,648 with mean value
    // 0 if the values are uniformly distributed. For a uniform distribution
    // over range [min, max], the standard deviation of a single value is
    // (max-min)/√12 = 1239850261. For a set of N of such values, the standard
    // deviation is smaller than that of an individual sample by a factor of
    // 1/√N, or 1239850.
    EXPECT_NEAR(0., rs.mean(), 2.*1239850.);

    // For a normal distribution the sample variance is σ^2. Let S^2 be the
    // set variance based on N samples. If N is very large, then S^2
    // approximately has a normal distribution with mean σ^2 and variance
    // 2σ^4/(n−1).
    const double sigma2 = 1239850261.*1239850261.;
    EXPECT_NEAR(sigma2, rs.variance(), 2*sqrt(2.*sigma2*sigma2/(N-1)));
}

TEST_F(RandomTest, nextIntRange)
{
    const uint32_t rangeMax = static_cast<uint32_t>(abs(random->nextInt()));
    set<int32_t> values;
    const int NITERATIONS = 65536;
    for (int i=0; i<NITERATIONS; ++i)
        values.insert(random->nextInt(rangeMax));
    EXPECT_LE(0, *values.begin());
    EXPECT_GE(rangeMax-1, (unsigned)(*values.rbegin()));
}

TEST_F(RandomTest, nextLongRange)
{
    const uint64_t rangeMax = static_cast<uint64_t>(abs(random->nextLong()));
    set<int64_t> values;
    const int NITERATIONS = 65536;
    for (int i=0; i<NITERATIONS; ++i)
        values.insert(random->nextLong(rangeMax));
    EXPECT_LE(0, *values.begin());
    EXPECT_GE(rangeMax-1, (unsigned)(*values.rbegin()));
}

}}} // namespace netflix::msl::crypto
